### ISSUE TITLE
lint against `todo!()` and `dbg!()`

### ISSUE THREAD

**Issue by cgwalters** [rust-lang/rust-clippy#9260]
### What it does

I heavily use the `todo!()` and `dbg!()` macros during development.  However, I think most projects never actually want to have these in production code committed to the repository.

In one project, I actually did commit `dbg!()` and was embarrassed to find it later; I added a manual check for `dbg!()`: https://github.com/ostreedev/ostree-rs-ext/pull/329

But, does everyone agree clippy should warn about these by default?

In the project, I already have a CI check that runs clippy that would have then errored out on this.

### Lint Name

todo-and-dbg

### Advantage

No one wants `todo!()` or `dbg!()` in code committed in "main" branches.

### Drawbacks

Some people may be using `dbg!()` in "production" code, but they can switch to e.g. `eprintln!("{:?}", val)` or any number of logging libraries.

Users of `todo!()` can switch to e.g. `unimplemented!()`.

### Example

```rust
todo!()
```

Could be written as:

```rust
# well, the real code here
```

**Comment by 5225225** [comment]
https://rust-lang.github.io/rust-clippy/master/index.html#todo and https://rust-lang.github.io/rust-clippy/master/index.html#dbg_macro both already exist, so I suppose the question is should they be on by default?

**Comment by matthiaskrgr** [comment]
I use `dbg!()` in tests and I find that very useful.
`dbg!()` and `eprintln!("{:?}", val)` are not the same, as dbg! will still return the value while eprintln!() won't so its much more versatile, you can't do this with eprintln!() for example:
````rust
fn a(truthiness: bool) {
    if dbg!(truthiness) {
        std::process::exit(1);
    }
}

fn main() {
    a(true);
}
````
output:
`[src/main.rs:2] truthiness = true`


`todo!()` is nice for indicating "I know I still have to add some code here, but I don't want a compiler error right now (think `fn sum_of_something() -> i32 { todo!() }`) and its also shorter to type.

There is nothing wrong with the macros per se so I would assume that people who write todo!() or dbg!() do this intentionally and would be annoyed by the lint, and those would who do not want todo or dbg in their code simply will not write it into their code, so this lint would almost always annoy the wrong users.

IMO this should stay in restriction.
The number of repos that have enabled clippy::todo (83) and clippy::dbg_macro (589) is negligible compared to the number of repos that have at least enabled extra clippy lints (39K), even in case of dbg!() that's ~1.2%  (source: github code search)

**Comment by cgwalters** [comment]
> https://rust-lang.github.io/rust-clippy/master/index.html#todo and https://rust-lang.github.io/rust-clippy/master/index.html#dbg_macro both already exist

I'm sorry I completely forgot there was a set of not-on-by-default lints.  I'll enable them in my repositories - sorry for the noise!

> so I suppose the question is should they be on by default?

I personally don't have a strong opinion.  As always with these things I think it's interesting to try a search:

- https://cs.github.com/?scopeName=All+repos&scope=&q=%22%23%21%5Bdeny%28clippy%3A%3Adbg_macro%29%5D%22
- https://sourcegraph.com/search?q=context:global+%23%21%5Bdeny%28clippy::dbg_macro%29%5D&patternType=literal

Not many hits.  But I'm going to add it in my repos, FWIW.

**Comment by cgwalters** [comment]
Eh, if we want to consider enabling them by default I think it should probably be a new issue, so closing.
